### DISCUSSION TITLE
oAuth 2 chaining using keycloak

### DISCUSSION THREAD

**Discussion by gauravji2192** [keycloak/keycloak#29196] [Q&A]
Hello community, i would like to do OAuth 2.0 chaining using keycloak for CIBA and Auth code flows. There are two scenarios:
a) CIBA flow: OAuth2 client makes CIBA call to Org A keycloak and Org A keycloak makes CIBA call to Org B keycloak. Org B keycloak is main identity provider and provides token. Org A keycloak is using store and map to a self generated reference token strategy. Organization B has final resources and need to verify identity of user and collect consent.
b) Auth codeflow : OAuth2 client makes Authcode call to Org A keycloak and Org A keycloak redirects OAuth2 client with necessary information to all to Org B keycloak. Org B keycloak is main identity provider. Org A keycloak is using store and map to a self-generated reference token strategy. Organization B has final resources and need to verify identity of user and collect consent.

For example,
OAuth2 client ---> Organization A Keycloak --> Organization B Keycloak

I am unable to find any such configuration example and information. Please help.

**Comment by codespearhead** [comment]
The OAuth2 client initiates a Client-Initiated Backchannel Authentication Flow ([CIBA](https://cloudentity.com/developers/basics/oauth-grant-types/ciba-flow/)) with Organization A's Keycloak. If successful, Organization A's Keycloak returns a [transaction identifier](https://curity.io/resources/learn/ciba-flow#:~:text=auth_req_id) (`auth_req_id`) to the client. Then the client uses this identifier to initiate an [Authorization Code Flow](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow) with Organization B's Keycloak server, which in turn issues the necessary access and/or ID tokens.

So there isn't much to it conceptually. Here's a revised [step-by-step guide on the implementation](https://github.com/keycloak/keycloak/files/15183058/discussion-29196.md) by [Google's Gemini](https://gemini.google.com/).

**Reply by gauravji2192** [reply]
Apologies, i did not explain my question well. There are two scenarios:

a) OAuth2 client makes CIBA call to Org A keycloak and Org A keycloak makes CIBA call to Org B keycloak. Org B keycloak is main identity provider and provides token. Org A keycloak is using store and map to a self generated reference token strategy.
b) OAuth2 client makes Authcode call to Org A keycloak and Org A keycloak redirects OAuth2 client with necessary information to all to Org B keycloak. Org B keycloak is main identity provider. Org A keycloak is using store and map to a self generated reference token strategy.

Can you help with this scenario?
